classes/local/store/azure_blob_storage/client.php

```diff
@@ -0,0 +1,296 @@
+
+use admin_settingpage;
+use coding_exception;
+use GuzzleHttp\Psr7\Utils;
+use tool_objectfs\local\store\object_client_base;
+use local_azureblobstorage\api;
+use local_azureblobstorage\stream_wrapper;
+use stdClass;
+use Throwable;
+
+/**
+ * Azure blob storage client
+ *
+ * @package    tool_objectfs
+ * @author     Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @copyright  2024 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class client extends object_client_base {
+
+    /** @var api $api Azure API */
+    protected api $api;
+
+    /**
+     * Creates object client
+     * @param stdClass $config / TODO is this maybe null ?
+     */
+    public function __construct($config) {
+        if (empty($config) || !$this->get_availability()) {
+            parent::__construct($config);
+            return;
+        }
+
+        $this->api = new api($config->azure_accountname, $config->azure_container, $config->azure_sastoken);
+        $this->config = $config;
+        $this->maxupload = api::MAX_BLOCK_SIZE;
+    }
+
+    /**
+     * Determines if this filesystem is available for use.
+     * @return bool
+     */
+    public function get_availability(): bool {
+        // Requires local_azureblobstorage to be installed.
+        // Namespace changed in 4.4+.
+        if (class_exists('\core\plugin_manager')) {
+            $info = \core\plugin_manager::instance()->get_plugin_info('local_azureblobstorage');
+            // For 4.2 and 4.3.
+        } else if (class_exists('\core_plugin_manager')) {
+            $info = \core_plugin_manager::instance()->get_plugin_info('local_azureblobstorage');
+        } else {
+            throw new coding_exception("Could not load plugin manager class");
+        }
+
+        // Info is empty if plugin is not installed or no API is setup (missing config?).
+        return !empty($info);
+    }
+
+    /**
+     * Sets the StreamWrapper to allow accessing the remote content via a blob:// path.
+     */
+    public function register_stream_wrapper() {
+        if ($this->get_availability()) {
+            stream_wrapper::register($this->api);
+        } else {
+            parent::register_stream_wrapper();
+        }
+    }
+
+    /**
+     * Returns the full path for a given file by contenthash
+     * @param string $contenthash
+     * @return string filepath
+     */
+    public function get_fullpath_from_hash($contenthash): string {
+        $filepath = $this->get_filepath_from_hash($contenthash);
+        return "blob://$filepath";
+    }
+
+    /**
+     * Returns the filepath from the contenthash, mimicking the
+     * structure of the filedir storage system.
+     * @param string $contenthash
+     * @return string filepath
+     */
+    protected function get_filepath_from_hash($contenthash): string {
+        $l1 = $contenthash[0] . $contenthash[1];
+        $l2 = $contenthash[2] . $contenthash[3];
+        return "$l1/$l2/$contenthash";
```

Comment:
Do we need to support ObjectFS bucket object prefixing here?

Reply:
just checked, bucket key prefix is a s3 client specific setting so doesn't apply here